File: src/main/java/hzt/aoc/day16/Part2TicketTranslation.java
package hzt.aoc.day16;

import hzt.collections.ListX;
import hzt.function.It;

import java.util.Arrays;
import java.util.List;

// credits to turkey dev
public class Part2TicketTranslation extends Day16Challenge {

    public Part2TicketTranslation() {
        super("part 2",
                "Once you work out which field is which, look for the six fields on your ticket that start with the word departure. \n" +
                        "What do you get if you multiply those six values together?");
    }

    private static final int FIRST_SIX_FIELDS = 6;

    @Override
    protected long solveByParsedInput(List<Field> fields, List<Integer> ourTicketValues, ListX<ListX<Integer>> nearbyTickets) {
        boolean[][] possibleMatches = new boolean[fields.size()][ourTicketValues.size()];
        for (boolean[] possibleMatch : possibleMatches) {
            Arrays.fill(possibleMatch, true);
        }
        var validTickets = findValidTickets(fields, nearbyTickets);
        for (var ticket : validTickets) {
            for (int col = 0; col < ticket.size(); col++) {
                for (int row = 0; row < fields.size(); row++) {
                    if (!fields.get(row).containsValueInRanges(ticket.get(col))) {
                        possibleMatches[row][col] = false;
                    }
                }
            }
        }
        LOGGER.trace(ourTicketValues);

        iterateUntilUniqueValueForeEachField(possibleMatches);
        if (LOGGER.isTraceEnabled()) {
            LOGGER.trace(booleanGrid2DAsString(possibleMatches));
        }
        return getAnswer(possibleMatches, ourTicketValues);
    }

    private void iterateUntilUniqueValueForeEachField(boolean[][] possibleMatches) {
        while (!isDone(possibleMatches)) {
            filterOutUniqueValues(possibleMatches);
        }
    }

    public void filterOutUniqueValues(boolean[][] possibleMatches) {
        for (int col = 0; col < possibleMatches[0].length; col++) {
            int count = 0;
            int index = -1;
            for (int row = 0; row < possibleMatches.length; row++) {
                if (possibleMatches[col][row]) {
                    count++;
                    index = row;
                }
            }
            if (count == 1) {
                for (int i = 0; i < possibleMatches.length; i++) {
                    if (i != col) {
                        possibleMatches[i][index] = false;
                    }
                }
            }
        }
    }

    public boolean isDone(boolean[][] possibleMatches) {
        for (boolean[] possibleMatch : possibleMatches) {
            int matches = ListX.ofBools(possibleMatch).count(It::self);
            if (matches > 1) {
                return false;
            }
        }
        return true;
    }

    protected List<ListX<Integer>> findValidTickets(List<Field> fields, ListX<ListX<Integer>> nearbyTickets) {
        return nearbyTickets.filterToMutableList(ticket -> ticket.all(value -> fieldsContainValue(value, fields)));
    }

    private long getAnswer(boolean[][] possibleMatches, List<Integer> ourTicketValues) {
        long answer = 1;
        for (int row = 0; row < FIRST_SIX_FIELDS; row++) {
            for (int col = 0; col < possibleMatches.length; col++) {
                if (possibleMatches[row][col]) {
                    int value = ourTicketValues.get(col);
                    answer *= value;
                    break;
                }
            }
        }
        return answer;
    }

    @Override
    String getMessage(long global) {
        return String.format("%d", global);
    }
}

File: src/main/kotlin/hzt/aoc/day16/Part2TicketTranslation.kt
package hzt.aoc.day16

import java.util.Arrays
import java.util.ArrayList

// credits to turkey dev
class Part2TicketTranslation : Day16Challenge(
    "part 2",
    """
         Once you work out which field is which, look for the six fields on your ticket that start with the word departure. 
         What do you get if you multiply those six values together
         """.trimIndent()
) {
    override fun solveByParsedInput(
        fields: List<Field>,
        yourTicketValues: List<Int>,
        nearbyTicketValues: List<List<Int>>
    ): Long {
        val possibleMatches = Array(fields.size) { BooleanArray(yourTicketValues.size) }
        for (possibleMatch in possibleMatches) Arrays.fill(possibleMatch, true)
        val validTickets = findValidTickets(fields, nearbyTicketValues)
        for (ticket in validTickets) {
            for (col in ticket.indices) {
                for (row in fields.indices) {
                    if (!fields[row].containsValueInRanges(ticket[col])) {
                        possibleMatches[row][col] = false
                    }
                }
            }
        }
        LOGGER.trace(yourTicketValues)
        iterateUntilUniqueValueForeEachField(possibleMatches)
        if (LOGGER.isTraceEnabled) LOGGER.trace(booleanGrid2DAsString(possibleMatches))
        return getAnswer(possibleMatches, yourTicketValues)
    }

    private fun iterateUntilUniqueValueForeEachField(possibleMatches: Array<BooleanArray>) {
        while (!isDone(possibleMatches)) filterOutUniqueValues(possibleMatches)
    }

    private fun filterOutUniqueValues(possibleMatches: Array<BooleanArray>) {
        for (col in 0 until possibleMatches[0].size) {
            var count = 0
            var index = -1
            for (row in possibleMatches.indices) {
                if (possibleMatches[col][row]) {
                    count++
                    index = row
                }
            }
            if (count == 1) {
                for (i in possibleMatches.indices) {
                    if (i != col) {
                        possibleMatches[i][index] = false
                    }
                }
            }
        }
    }

    private fun isDone(possibleMatches: Array<BooleanArray>): Boolean = possibleMatches
        .map { matches -> matches.count { it } }
        .none { it > 1 }

    private fun findValidTickets(fields: List<Field>, nearbyTickets: List<List<Int>>): List<List<Int>> {
        val validTickets: MutableList<List<Int>> = ArrayList()
        for (nearbyTicket in nearbyTickets) {
            val containsOnlyValidValues = nearbyTicket.all { fieldsContainValue(it, fields) }
            if (containsOnlyValidValues) validTickets.add(nearbyTicket)
        }
        return validTickets
    }

    private fun getAnswer(possibleMatches: Array<BooleanArray>, ourTicketValues: List<Int>): Long {
        var answer: Long = 1
        for (row in 0 until FIRST_SIX_FIELDS) {
            for (col in possibleMatches.indices) {
                if (possibleMatches[row][col]) {
                    val value = ourTicketValues[col]
                    answer *= value.toLong()
                    break
                }
            }
        }
        return answer
    }

    override fun getMessage(value: Long): String = String.format("%d", value)

    companion object {
        private const val FIRST_SIX_FIELDS = 6
    }
}
